### PR TITLE
Add bounty mechanics to coffeeconomy.

### PR DESCRIPTION
**Pull request by bzmwillemsen**
`.coffee bounty x` will start a bounty of `x` points
`.coffee bounty x` where a bounty already exists overrides the existing bounty only if the new price is more
`.brewing` will claim the bounty and dibs a spot for the bounty hunter
Points will be rewarded when the brew is ended.

### CODE REVIEW

File: src/coffee.coffee [hunk 1]
```diff
@@ -115,6 +119,25 @@ module.exports = (robot) ->
 
     createBrew msg
 
+  robot.respond /coffee bounty(?:\s+(\d+))?$/i, (msg) ->
+    if isBrewing()
+      msg.send "A brew is already underway by (@#{brewing.barista})!" +
+          "#{statusEmoji.randim('failure')}\n" + freeSpots()
```
Comment (hoylemd) [inline src/coffee.coffee]: `randim`?

File: src/coffee.coffee [hunk 2]
```diff
@@ -115,6 +119,25 @@ module.exports = (robot) ->
...
+
+    hunter = robot.brain.usersForFuzzyName(msg.message.user['name'])[0].name
```
Comment (hoylemd) [inline src/coffee.coffee]: `robot.brain` hahahahahahaha

File: src/coffee.coffee [hunk 3]
```diff
@@ -160,12 +185,24 @@ module.exports = (robot) ->
...
+    if bounty.amount > 0 and barista is not bounty.hunter
+      msg.send "@team A bountied Brew has been started by @#{brewing.barista}! #{statusEmoji.random('success')}\n" +
+          "@#{brewing.barista} will get an extra #{bounty.amount} coffeeconomy points for this brew."
+          "To grab a spot use: #{robot.alias}dibs\n" +
+          "To end the brew use: #{robot.alias}fresh pot"
+
+      dibs(bounty.hunter)
+    if bounty.amount > 0 and barista is bounty.hunter
+      msg.send "Hey @{bounty.hunter} can't collect your own bounty! Your bounty has been cancelled."
+      bounty.amount = 0
```
Comment (hoylemd) [inline src/coffee.coffee]: This should probably also clear the bounty 'hunter', or it's going to behave like they still have a bounty of 0 out.

File: src/coffee.coffee [hunk 4]
```diff
@@ -181,6 +218,12 @@ module.exports = (robot) ->
 
         if dibber == brewing.barista
           coffeeconomy.deposit client, dibsLimit - 1
+
+          if bounty.price > 0
+            coffeeconomy.deposit client, bounty.price
```
Comment (hoylemd) [inline src/coffee.coffee]: isn't it `bounty.amount'?

**Comment by hoylemd** [comment]
first review pass done. some naming issues.

File: src/coffee.coffee [hunk 5]
```diff
@@ -115,6 +119,37 @@ module.exports = (robot) ->
 
     createBrew msg
 
+  robot.respond /coffee bounty\s+(\d+)$/i, (msg) ->
+    if isBrewing()
+      msg.send "A brew is already underway by (@#{brewing.barista})!" +
+          "#{statusEmoji.random('failure')}\n" + freeSpots()
```
Comment (brentc) [inline src/coffee.coffee]: Add a `return` to fail out

File: src/coffee.coffee [hunk 6]
```diff
@@ -115,6 +119,37 @@ module.exports = (robot) ->
...
+
+    issuer = robot.brain.usersForFuzzyName(msg.message.user['name'])[0].name
+    proposedPrice = parseInt msg.match[1]?.trim() || 0
```
Comment (brentc) [inline src/coffee.coffee]: s/Price/Reward/
s/price/reward/

File: src/coffee.coffee [hunk 7]
```diff
@@ -115,6 +119,37 @@ module.exports = (robot) ->
...
+    if hasBounty() and proposedPrice > bounty.price
+      if issuer is bounty.issuer
+        msg.send "@#{issuer} has raised the bounty from #{bounty.price} to #{proposedPrice}."
+      else
+        msg.send "@#{issuer} has outbid @#{bounty.issuer}'s bounty of #{bounty.price} by increasing it to #{proposedPrice}."
+    else if hasBounty() and proposedPrice <= bounty.price
+      msg.send "A bounty of #{bounty.price} by #{issuer} has already been put out. #{statusEmoji.random('failure')}"
+      return
+    else
+      msg.send "@team #{issuer} has put out a bounty of #{proposedPrice}."
```
Comment (brentc) [inline src/coffee.coffee]: "...has issued a bounty of..."

File: src/coffee.coffee [hunk 8]
```diff
@@ -124,7 +159,20 @@ module.exports = (robot) ->
     return handleNoBrew msg if not isBrewing()
 
     dibber = robot.brain.usersForFuzzyName(msg.message.user['name'])[0].name
+    dibs dibber, msg
```
Comment (brentc) [inline src/coffee.coffee]: Rename `dibs` to `dib`

File: src/coffee.coffee [hunk 9]
```diff
@@ -135,8 +183,12 @@ module.exports = (robot) ->
 
     brewing.dibs.push(dibber)
 
-    threadedMsg(msg).send "Ok @#{dibber}, you grabbed :coffee: ##{brewing.dibs.length}! " +
-        "#{statusEmoji.random('success')}"
+    if bounty.issuer is dibber and hasBounty()
```
Comment (brentc) [inline src/coffee.coffee]: Check `hasBounty()` first

File: src/coffee.coffee [hunk 10]
```diff
@@ -156,16 +208,30 @@ module.exports = (robot) ->
 
   createBrew = (msg) ->
     barista = robot.brain.usersForFuzzyName(msg.message.user['name'])[0].name
+
+    if bounty.price > 0 and barista is bounty.issuer
+      msg.send "Hey @#{bounty.issuer} can't collect your own bounty! Your bounty has been cancelled. #{statusEmoji.random('failure')}"
```
Comment (brentc) [inline src/coffee.coffee]: "Hey Brett can't collect your own" →  "Hey Brett, _you_ can't collect your own"

Comment (brentc) [inline src/coffee.coffee]: And wrap the "Your bounty has been cancelled" with "**"s

File: src/coffee.coffee [hunk 11]
```diff
@@ -156,16 +208,30 @@ module.exports = (robot) ->
...
+      msg.send "@team A bountied Brew has been started by @#{brewing.barista}! #{statusEmoji.random('success')}\n" +
+          "@#{brewing.barista} will get an extra #{bounty.price} :coffee:s for this brew.\n" +
+          "**To grab a spot use: #{robot.alias}dibs**\n" +
+          "**To end the brew use: #{robot.alias}fresh pot**"
+
+      dibs bounty.issuer, msg
+    else
+      msg.send "@team Brew started by @#{brewing.barista}! #{statusEmoji.random('success')}\n" +
+          "To grab a spot use: #{robot.alias}dibs\n" +
+          "To end the brew use: #{robot.alias}fresh pot"
```
Comment (brentc) [inline src/coffee.coffee]: This second version of the brew started doesn't bold the commands. 

Also while we're here let's change the formatting to:
```coffee
"To grab a spot use: `#{robot.alias}dibs`\n" +
```

No bolding

File: src/coffee.coffee [hunk 12]
```diff
@@ -223,7 +222,7 @@ module.exports = (robot) ->
 
     if hasBounty()
       msg.send "@team A bountied Brew has been started by @#{brewing.barista}! #{statusEmoji.random('success')}\n" +
-          "@#{brewing.barista} will get an extra #{bounty.price} :coffee:s for this brew.\n" +
+          "@#{brewing.barista} will get an extra #{bounty.reward} :coffee:s for this brew.\n" +
```
Comment (brentc) [inline src/coffee.coffee]: "will be rewarded #{bounty.reward}..."

File: src/coffee.coffee (hunk 5 above)
Comment (bzmwillemsen) [inline src/coffee.coffee]: Done.

File: src/coffee.coffee (hunk 6 above)
Comment (bzmwillemsen) [inline src/coffee.coffee]: Done.

File: src/coffee.coffee (hunk 7 above)
Comment (bzmwillemsen) [inline src/coffee.coffee]: Done.

File: src/coffee.coffee (hunk 8 above)
Comment (bzmwillemsen) [inline src/coffee.coffee]: Done.

File: src/coffee.coffee (hunk 9 above)
Comment (bzmwillemsen) [inline src/coffee.coffee]: Done.

File: src/coffee.coffee (hunk 10 above)
Comment (bzmwillemsen) [inline src/coffee.coffee]: Done.

File: src/coffee.coffee (hunk 11 above)
Comment (bzmwillemsen) [inline src/coffee.coffee]: Done.

File: src/coffee.coffee (hunk 12 above)
Comment (bzmwillemsen) [inline src/coffee.coffee]: Done.

**Comment by brentc** [comment]
@bzmwillemsen document the new command in the top of the file

File: src/coffee.coffee [hunk 13]
```diff
@@ -222,15 +222,15 @@ module.exports = (robot) ->
 
     if hasBounty()
       msg.send "@team A bountied Brew has been started by @#{brewing.barista}! #{statusEmoji.random('success')}\n" +
```
Comment (brentc) [inline src/coffee.coffee]: "A bountied brew"

Comment (bzmwillemsen) [inline src/coffee.coffee]: Done.

File: src/coffee.coffee [hunk 14]
```diff
@@ -156,16 +208,30 @@ module.exports = (robot) ->
 
   createBrew = (msg) ->
     barista = robot.brain.usersForFuzzyName(msg.message.user['name'])[0].name
+
+    if bounty.reward > 0 and barista is bounty.issuer
+      msg.send "Hey @#{bounty.issuer} you can't collect your own bounty! **Your bounty has been cancelled.** #{statusEmoji.random('failure')}"
+      clearBounty
+      return
```
Comment (brentc) [inline src/coffee.coffee]: Hmmm, should we `return` here? Or let the brewing command continue?

**Comment by brentc** [comment]
:+1:

File: src/coffee.coffee (hunk 12 above)
Comment (hoylemd) [inline src/coffee.coffee]: wouldn't 'awarded' be more correct?

**Pull request merged**